uniconf/daemon/uniconfd.8.in
'\" t
.\" Worldvisions Weaver Software
.\"   Copyright (C) 1997 - 2004 Net Integration Technologies Inc.
.TH UNICONFD 8 "August 2004" "UniConfDaemon #VERSION#"
.SH NAME
uniconfd \- a daemon program for the UniConf configuration system
.SH SYNOPSIS
.B uniconfd
[
.IR OPTIONS
]
.IR MOUNT\  ...
.SH DESCRIPTION
UniConf is the One True Configuration system that includes all the
others because it has plugin backends
.I and
frontends. Or, less grandiosely, it's a lightweight, distributed,
cacheable tree of strings.

It supports:
.IP \(bu 4
retrieving, storing, and enumerating key/value pairs (where both keys
and values are strings).
.IP \(bu
multiple backends where the actual key/value pairs are stored.
.IP \(bu
multiple frontends for tying it to other configuration architectures.
.PP
It operates locally, and across a network, allowing you to tie
multiple different applications together for distributed computing.
Also, it provides notifications in the form of callbacks, so your
application can be notified if a configuration key has changed.

.B uniconfd
is necessary when you have more than one application, or multiple
instances of an application, sharing one configuration.
UniConf-enabled applications contact
.B uniconfd
which provides notifications when any of their watched keys change.

You tell
.B uniconfd
which UniConf
.I MOUNT
you want it to manage.  See the
.B MOUNTS
section for more information.
.SH OPTIONS
.TP
.B \-f
Run in the foreground.  Do not fork into a separate daemon process.
.TP
.BR \-d ,\  \-dd
Print debugging messages to the console.  The second
.B d
increases the verbosity of the messages.
.TP
.B \-V
Print the version number and exit.
.TP
.B \-a
Require authentication on incoming connections.
.TP
.B \-A
Check all accesses against a
.I perms
moniker.
.TP
.BI \-p\  port
Listen on a given TCP 
.IR port .
The default is 4111.  If
.I port
is 0, then listening on TCP is disabled.
.TP
.BI \-s\  port
Listen on a given TCP
.I port
wrapped in SSL.  
The default is 4112.  If
.I port
is 0, then listening on SSL-over-TCP is disabled.
.TP
.BI \-u\  filename
Listen on a given Unix socket
.IR filename .
This is disabled by default.
.SH MOUNTS
Mounts are UniConf path monikers which are in the form:
.RS
.RI / SUBTREE = GENERATORS : PATH
.RE
.TP
.I SUBTREE
This is the tree to manage.  All trees are descended from the root
tree, indicated by a bare slash 
.RB ( / ).
.TP
.I GENERATORS
These are the generators used to read and write key/value pairs.  You
can chain them with colons.  For example, the generator chain:
.I cache:retry:ini
will cache the configuration for speed, retry persistently if the data
source disappears, and store the data in an INI-formatted file.
.TP
.I PATH
This is the location where the data is stored.  It is dependent on
which
.IR GENERATOR S
were specified.  For instance, it could be:
\(bu
a filename
.RI (ini: /var/lib/app/config.ini ),
.br
\(bu
a network address,
.RI (tcp: open.nit.ca:4111 ),
.br
\(bu
or even an empty string
.R (tmp:).
.PP
.B Examples:
.RS
.RI /= tmp :
.br
.RI / ca/nit = ssl : open.nit.ca
.br
.RI / ca/nit/uniconfd = ini : /var/lib/uniconfd/uniconfd.ini
.br
.RI / apps = cache:retry:unix : /var/lib/apps/socket
.RE
.SH FILES
.I /etc/uniconfd.conf
.br
.I /var/lib/uniconf/uniconfd.ini
.br
.I /var/lib/uniconf/uniconf.ini
.SH AUTHORS
This software was written by the hackers at Net Integration
Technologies.  Contact us at <wvstreams\-dev@lists.nit.ca>
.\" Local variables:
.\" mode:nroff
.\" End:


CHANGES
Changes to WvStreams since original release:

3.69f (2002-02-06)

- None



uniconf/daemon/uniconfd.cc
#include "wvautoconf.h"
#ifdef HAVE_UNISTD_H
# include <unistd.h>
#endif
#ifdef HAVE_GETOPT_H
# include <getopt.h>
#endif

#ifndef _WIN32
#include <signal.h>
#include <sys/types.h>
#include <sys/stat.h>
#endif

#ifdef WITH_SLP
#include "wvslp.h"
#endif

#include "wvlogrcv.h"
#include "uniconfdaemon.h"
#include "uniclientconn.h"
#include "unisecuregen.h"
#include "unipermgen.h"
#include "uniconfroot.h"
#include "wvstrutils.h"
#include "wvfileutils.h"
#include "wvstreamsdaemon.h"
#include "wvlinkerhack.h"

#ifdef WITH_SLP
#include "slp.h"
#endif

#include <map>

using std::map;
using wv::shared_ptr;

WV_LINK_TO(UniGenHack);

#define DEFAULT_CONFIG_FILE "ini:uniconf.ini"


static map<WvString, shared_ptr<IUniConfGen> > namedgens;


static IUniConfGen *creator(WvStringParm s, IObject*)
{
    map<WvString, shared_ptr<IUniConfGen> >::iterator it = namedgens.find(s);
    shared_ptr<IUniConfGen> gen;

    if (it != namedgens.end())
	gen = it->second;

    if (gen)
	gen->addRef();

    return gen.get();
}

WvMoniker<IUniConfGen> UniNamedMoniker("named", creator);


class UniConfd : public WvStreamsDaemon
{
    bool needauth;
    WvString permmon;
    WvStringList lmonikers;
    time_t commit_interval;

    UniConfRoot cfg;
    bool first_time;
    IUniConfGen *permgen;
    
    bool namedgen_cb(WvStringParm option, void *)
    {
	WvString name(option);
	WvString moniker;
	char* ptr;

	ptr = strchr(name.edit(), '=');

	if (!ptr)
	    return false;

	*ptr = 0;
	moniker = ptr + 1;

	namedgens[name] = shared_ptr<IUniConfGen>(
	    wvcreate<IUniConfGen>(moniker),
	    wv::bind(&IUniConfGen::release, _1));

	return true;
    }

    void commit_stream_cb(WvStream *s)
    {
	cfg.commit();
	cfg.refresh();
	if (permgen)
	    permgen->refresh();
	    
        s->alarm(commit_interval * 1000);
    }
    
    void startup()
    {
        if (first_time)
        {
            WvStringList::Iter i(_extra_args);
            for (i.rewind(); i.next(); )
            {
	        WvString path = *i, moniker;
	        char *cptr = strchr(path.edit(), '=');
	        if (!cptr)
	        {
	            moniker = path;
	            path = "/";
	        }
	        else
	        {
	            *cptr = 0;
	            moniker = cptr+1;
	        }
	        
	        log("Mounting '%s' on '%s': ", moniker, path);
	        IUniConfGen *gen = cfg[path].mount(moniker, false);
	        if (gen && gen->isok())
	            log("ok.\n");
	        else
	            log("FAILED!\n");
            }
            
            cfg.refresh();
        }

        permgen = !!permmon ? wvcreate<IUniConfGen>(permmon) : NULL;
        
        UniConfDaemon *daemon = new UniConfDaemon(cfg, needauth, permgen);
        add_die_stream(daemon, true, "uniconfd");
	
	if (lmonikers.isempty())
	{
	    log(WvLog::Critical, "Can't start: no listeners given!\n");
	    die(7);
	    return;
	}

	WvStringList::Iter i(lmonikers);
	for (i.rewind(); i.next(); )
	    daemon->listen(*i);

        WvStream *commit_stream = new WvStream;
        commit_stream->setcallback(wv::bind(&UniConfd::commit_stream_cb, this,
					    commit_stream));
        commit_stream->alarm(commit_interval * 1000);
        add_die_stream(commit_stream, true, "commit");
        
        if (first_time)
            first_time = false;
    }
    
public:

    UniConfd():
	WvStreamsDaemon("uniconfd", VERBOSE_WVPACKAGE_VERSION,
			wv::bind(&UniConfd::startup, this)),
	needauth(false),
	commit_interval(5*60),
	first_time(true),
	permgen(NULL)
    {
        args.add_option(0, "pid-file",
                "Specify the .pid file to use (only applies with --daemonize)", "filename",
                pid_file);
        args.add_set_bool_option('a', "need-auth",
                "Require authentication on incoming connections", needauth);
        args.add_option('A', "check-access",
                "Check all accesses against perms moniker", "moniker",
                permmon);
	args.add_option('l', "listen",
		"Listen on the given socket (eg. tcp:4111, ssl:tcp:4112)",
		"lmoniker", lmonikers);
	args.add_option('n', "named-gen",
			"creates a \"named\" moniker 'name' from 'moniker'",
			"name=moniker",
			wv::bind(&UniConfd::namedgen_cb, this, _1, _2), NULL);
	args.add_optional_arg("MONIKERS", true);
	args.set_email("<" WVPACKAGE_BUGREPORT ">");
    }
    
    
};

int main(int argc, char **argv)
{
    UniConfd uniconfd;
   
    return uniconfd.run(argc, argv);
}
